plugins/inputs/p4runtime/p4runtime.go

```diff
@@ -0,0 +1,232 @@
+			defer p.wg.Done()
+			entries, err := p.readAllEntries(counter.Preamble.Id)
+			if err != nil {
+				acc.AddError(
+					fmt.Errorf(
+						"Reading counter entries with ID=%v failed with error: %v",
+						counter.Preamble.Id,
+						err,
+					),
+				)
+				return
+			}
+
+			for _, entry := range entries {
+				ce := entry.GetCounterEntry()
+
+				if ce == nil {
+					acc.AddError(fmt.Errorf("Reading counter entry from entry %v failed", entry))
+					continue
+				}
+
+				if ce.Data.ByteCount == 0 && ce.Data.PacketCount == 0 {
+					continue
+				}
+
+				fields := map[string]interface{}{
+					"bytes":         ce.Data.ByteCount,
+					"packets":       ce.Data.PacketCount,
+					"counter_index": ce.Index.Index,
+				}
+
+				tags := map[string]string{
+					"p4program_name": p4Info.PkgInfo.Name,
+					"counter_name":   counter.Preamble.Name,
+					"counter_type":   counter.Spec.Unit.String(),
+				}
+
+				acc.AddFields("p4_runtime", fields, tags)
+			}
+		}(counter)
+	}
+	p.wg.Wait()
+	return nil
+}
+
+func (p *P4runtime) Stop() {
+	p.conn.Close()
+	p.wg.Wait()
+}
+
+func initConnection(endpoint string, tlscfg *tls.Config) (*grpc.ClientConn, error) {
+	var creds credentials.TransportCredentials
+	if tlscfg != nil {
+		creds = credentials.NewTLS(tlscfg)
+	} else {
+		creds = insecure.NewCredentials()
+	}
+	return grpc.Dial(endpoint, grpc.WithTransportCredentials(creds))
+}
+
+func (p *P4runtime) getP4Info() (*p4ConfigV1.P4Info, error) {
+	req := &p4v1.GetForwardingPipelineConfigRequest{
+		DeviceId:     p.DeviceID,
+		ResponseType: p4v1.GetForwardingPipelineConfigRequest_ALL,
+	}
+	resp, err := p.client.GetForwardingPipelineConfig(context.Background(), req)
+	if err != nil {
+		return nil, fmt.Errorf("error when retrieving forwarding pipeline config: %v", err)
+	}
+
+	config := resp.GetConfig()
+	if config == nil {
+		return nil, fmt.Errorf(
+			"error when retrieving config from forwarding pipeline - pipeline doesn't have a config yet: %v",
+			err,
+		)
+	}
+
+	p4info := config.GetP4Info()
+	if p4info == nil {
+		return nil, fmt.Errorf(
+			"error when retrieving P4Info from config - config doesn't have a P4Info: %v",
+			err,
+		)
+	}
+
+	return p4info, nil
+}
+
+func filterCounters(counters []*p4ConfigV1.Counter, counterNamesInclude []string) []*p4ConfigV1.Counter {
+	if len(counterNamesInclude) == 0 {
+		return counters
+	}
+
+	var filteredCounters []*p4ConfigV1.Counter
+	for _, counter := range counters {
+		if counter == nil {
+			continue
+		}
+		if slices.Contains(counterNamesInclude, counter.Preamble.Name) {
+			filteredCounters = append(filteredCounters, counter)
+		}
+	}
+	return filteredCounters
+}
+
+func (p *P4runtime) newP4RuntimeClient() error {
+	var tlscfg *tls.Config
+	var err error
+
+	if p.EnableTLS {
+		if tlscfg, err = p.ClientConfig.TLSConfig(); err != nil {
+			return err
+		}
+	}
+
+	conn, err := initConnection(p.Endpoint, tlscfg)
+	if err != nil {
+		return fmt.Errorf("cannot connect to the server: %v", err)
```

Comment:
Can you please always wrap the error so we can later check for its type upstream...
```suggestion
		return fmt.Errorf("cannot connect to the server: %w", err)
```
Same for other instances in the code.

Reply:
replaced in all occurrences